### PR TITLE
useObserver

### LINKED ISSUE #23: Idea: Add useObserver
```js
function MyObserverComponent(props) {
  const observe = useObserver();

  return observe(() => {
    return <div>props.observable.x</div>
  });
}
```

I think it could have several advantages:
1. you can specify the exact specific piece of code that runs inside the reaction tracking, so you can do observable mutations outside it easily without triggering unwanted renders
2. you can memo your component, forward ref it or whatever you want without observer getting in the middle

Maybe the only disadvantage though is that you actually have to write "observe" or whatever name you assigned over the render phase though

Maybe offer it as an alternative to "observer" at least for advance use cases?

observer uses something like that internally anyway

_Originally posted by @xaviergonz in https://github.com/mobxjs/mobx-react-lite/issues/7#issuecomment-439679656_

### PR DESCRIPTION
**Pull request by xaviergonz**
Closes #23

### CODE REVIEW

File: src/observer.ts [hunk 1]
```diff
@@ -81,7 +72,7 @@ function useForceUpdate() {
     return update
 }
 
-function useObserverReaction(baseComponentName: string) {
+export function useObserver<T>(fn: () => T, baseComponentName = "anonymous"): T {
```
Comment (FredyC) [inline src/observer.ts]: Extract it to a separate file, please. And some basic tests would be nice as well :)

File: README.md [hunk 2]
```diff
@@ -112,6 +113,28 @@ function ObservePerson(props) {
 }
 ```
 
+### `useObserver<T>(fn: () => T, baseComponentName = "anonymous"): T`
+
+Low level implementation used internally by `observer`.
+It allows you to use an `observer` like behaviour, but still allowing you to optimize the component in any way you want (e.g. using `memo` with a custom
+`areEqual`, using `forwardRef`, etc.) and to declare exactly the part that is observed (the render phase). One good thing about this is that if any
```
Comment (FredyC) [inline README.md]: nit ... remove line break here and the line above, please :)

File: README.md [hunk 3]
```diff
@@ -112,6 +113,28 @@ function ObservePerson(props) {
...
+hook changes an observable for some reason then the component won't rerender twice unnecessarily.
+
+```
```
Comment (FredyC) [inline README.md]:
```suggestion
```tsx
```

File: README.md (hunk 2 above)
Comment (xaviergonz) [reply README.md]: done

File: src/observer.ts (hunk 1 above)
Comment (xaviergonz) [reply src/observer.ts]: done, the tests are now the same ones than for observer (except for one case I think)

File: test/observer.test.tsx [hunk 4]
```diff
@@ -2,412 +2,401 @@ import * as mobx from "mobx"
...
-                    title: "a"
-                }
-            ]
-        })
+function runTestSuite(mode: "observer" | "useObserver") {
+    function obsComponent<P extends object>(component: React.FunctionComponent<P>) {
+        if (mode === "observer") {
+            return observer(component)
+        } else {
+            return React.memo((props: P) => {
```
[the remark this answers is not in the document]
Comment (xaviergonz) [reply test/observer.test.tsx]: done

**Comment by FredyC** [comment]
Thanks once again :)

**Comment by xaviergonz** [comment]
no problem :)

**Comment by FredyC** [comment]
Released in [v0.3.0](https://github.com/mobxjs/mobx-react-lite/releases/tag/v0.3.0)

**Pull request merged**